{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "bfa05e6e-ace9-4569-819e-c4b133b13590",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sample to be processed\n",
    "study = 'scmznos_valtor_combined'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "87a8ef3c-7d63-406c-a2e5-9eb9d7049db0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input/output/resource directories\n",
    "place = 'nemo'\n",
    "scratchdir = \"/plus/scratch/users/scott/projects/scmznos_valtor\"\n",
    "rawdir = f'{scratchdir}/raw_data'\n",
    "outdir = f'{scratchdir}/project_results'\n",
    "datadir = f'{scratchdir}/resources'\n",
    "commonsdir = '/plus/data/@data_scott/common_resources'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "5339940f-f72a-45df-a56a-7940f693cd69",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load libraries\n",
    "import os\n",
    "import sklearn\n",
    "import pickle\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy\n",
    "from scipy.sparse import csr_matrix\n",
    "from scipy import io\n",
    "\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import pybiomart as pbm\n",
    "import leidenalg as la\n",
    "import scrublet as scr\n",
    "\n",
    "import graphtools as gt\n",
    "from pygsp import graphs, filters\n",
    "import phate\n",
    "import magic\n",
    "import scprep\n",
    "import sklearn\n",
    "import meld\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import seaborn as sns\n",
    "import time\n",
    "import natsort\n",
    "\n",
    "# Package settings\n",
    "sc.settings.autosave = False\n",
    "sc.settings.figdir = f'{outdir}/'\n",
    "np.random.seed(42)\n",
    "font = {'size'   : 14}\n",
    "mpl.rc('font', **font)\n",
    "mpl.rcParams['animation.embed_limit'] = 1000\n",
    "mpl.rcParams['pdf.fonttype'] = 42"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "160e0efa-2b36-4eb4-81da-1f25b6311542",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import sequence analysis tools\n",
    "import pyranges as pr\n",
    "import pybedtools\n",
    "import Bio\n",
    "import pysam\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "673b6511-0750-4188-9884-27978f9498e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<details>\n",
       "<summary>Click to view session information</summary>\n",
       "<pre>\n",
       "-----\n",
       "Bio                 1.81\n",
       "anndata             0.9.1\n",
       "graphtools          1.5.3\n",
       "leidenalg           0.10.0\n",
       "magic               3.0.0\n",
       "matplotlib          3.7.2\n",
       "meld                1.0.0\n",
       "natsort             8.4.0\n",
       "numpy               1.24.4\n",
       "pandas              2.2.1\n",
       "phate               1.0.11\n",
       "polars              0.18.15\n",
       "pyFunctions         NA\n",
       "pybedtools          0.9.1\n",
       "pybiomart           0.2.0\n",
       "pygsp               0.5.1\n",
       "pyranges            0.0.127\n",
       "pysam               0.21.0\n",
       "scanpy              1.9.3\n",
       "scipy               1.11.1\n",
       "scprep              1.2.3\n",
       "scrublet            NA\n",
       "seaborn             0.12.2\n",
       "session_info        1.0.0\n",
       "sklearn             1.3.0\n",
       "-----\n",
       "</pre>\n",
       "<details>\n",
       "<summary>Click to view modules imported as dependencies</summary>\n",
       "<pre>\n",
       "PIL                 10.0.0\n",
       "asttokens           NA\n",
       "attr                23.1.0\n",
       "backcall            0.2.0\n",
       "cattr               NA\n",
       "cattrs              NA\n",
       "certifi             2023.05.07\n",
       "cffi                1.15.1\n",
       "charset_normalizer  3.2.0\n",
       "cloudpickle         2.2.1\n",
       "comm                0.1.3\n",
       "cycler              0.10.0\n",
       "cython_runtime      NA\n",
       "cytoolz             0.12.2\n",
       "dask                2023.8.0\n",
       "dateutil            2.8.2\n",
       "debugpy             1.6.7\n",
       "decorator           5.1.1\n",
       "defusedxml          0.7.1\n",
       "deprecated          1.2.14\n",
       "executing           1.2.0\n",
       "fisher              v0.1.10\n",
       "future              0.18.3\n",
       "google              NA\n",
       "h5py                3.9.0\n",
       "idna                3.4\n",
       "igraph              0.10.6\n",
       "importlib_metadata  NA\n",
       "ipykernel           6.24.0\n",
       "ipywidgets          8.1.0\n",
       "jedi                0.18.2\n",
       "jinja2              3.1.2\n",
       "joblib              1.3.1\n",
       "kiwisolver          1.4.4\n",
       "llvmlite            0.40.1\n",
       "lz4                 4.3.2\n",
       "markupsafe          2.1.3\n",
       "mpl_toolkits        NA\n",
       "msgpack             1.0.5\n",
       "ncls                0.0.68\n",
       "numba               0.57.1\n",
       "numexpr             2.8.5\n",
       "orjson              3.9.5\n",
       "packaging           23.1\n",
       "parso               0.8.3\n",
       "patsy               0.5.3\n",
       "pexpect             4.8.0\n",
       "pickleshare         0.7.5\n",
       "pkg_resources       NA\n",
       "platformdirs        3.8.1\n",
       "plotly              5.19.0\n",
       "prompt_toolkit      3.0.39\n",
       "psutil              5.9.5\n",
       "ptyprocess          0.7.0\n",
       "pure_eval           0.2.2\n",
       "pyarrow             12.0.1\n",
       "pydev_ipython       NA\n",
       "pydevconsole        NA\n",
       "pydevd              2.9.5\n",
       "pydevd_file_utils   NA\n",
       "pydevd_plugins      NA\n",
       "pydevd_tracing      NA\n",
       "pygments            2.15.1\n",
       "pyparsing           3.0.9\n",
       "pytz                2023.3\n",
       "requests            2.31.0\n",
       "requests_cache      1.1.0\n",
       "s_gd2               1.8.1\n",
       "setuptools_scm      NA\n",
       "six                 1.16.0\n",
       "sorted_nearest      0.0.39\n",
       "sphinxcontrib       NA\n",
       "stack_data          0.6.2\n",
       "statsmodels         0.14.0\n",
       "swig_runtime_data4  NA\n",
       "tasklogger          1.2.0\n",
       "tblib               2.0.0\n",
       "texttable           1.6.7\n",
       "threadpoolctl       3.2.0\n",
       "tlz                 0.12.2\n",
       "toolz               0.12.0\n",
       "tornado             6.3.2\n",
       "traitlets           5.9.0\n",
       "typing_extensions   NA\n",
       "url_normalize       1.4.3\n",
       "urllib3             1.26.16\n",
       "wcwidth             0.2.6\n",
       "wrapt               1.15.0\n",
       "xxhash              NA\n",
       "yaml                6.0.1\n",
       "zipp                NA\n",
       "zmq                 25.1.0\n",
       "zstandard           0.21.0\n",
       "</pre>\n",
       "</details> <!-- seems like this ends pre, so might as well be explicit -->\n",
       "<pre>\n",
       "-----\n",
       "IPython             8.14.0\n",
       "jupyter_client      8.3.0\n",
       "jupyter_core        5.3.1\n",
       "-----\n",
       "Python 3.11.6 (main, Nov 14 2023, 09:36:21) [GCC 13.2.1 20230801]\n",
       "Linux-6.6.9-arch1-1-x86_64-with-glibc2.38\n",
       "-----\n",
       "Session information updated at 2024-04-16 11:52\n",
       "</pre>\n",
       "</details>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show loaded libraries\n",
    "import session_info\n",
    "session_info.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "528f4e59-bf62-4e09-869d-5902cda06cf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Homemade utility functions\n",
    "import pyFunctions\n",
    "from pyFunctions.annotation import *\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "6492d7e3-c333-4c80-8353-701c5870cf4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Load preprocessed data\n",
    "# sdata = sc.read_h5ad(f'{scratchdir}/preprocessed_data/Anndata_{study}_cells_preprocessed.h5ad')\n",
    "# sdata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e47fbdd7-9b42-41aa-8a77-b5f8b389995f",
   "metadata": {},
   "source": [
    "### Find genes containing mir430 binding sites in their 3'UTR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "06191ddd-f6a0-4f26-aebd-62e9a0a651dc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/scott/scripts/scmultiome_curbos_scripts/multiome/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n",
      "/home/scott/scripts/scmultiome_curbos_scripts/multiome/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_version</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_source</th>\n",
       "      <th>transcript_biotype</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_version</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>protein_version</th>\n",
       "      <th>tag</th>\n",
       "      <th>interval</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>27977296</td>\n",
       "      <td>28020042</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSDARG00000100083</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1:27977296-28020042</td>\n",
       "      <td>42746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>27984392</td>\n",
       "      <td>27995611</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSDARG00000100083</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>havana</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1:27984392-27995611</td>\n",
       "      <td>11219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>27984392</td>\n",
       "      <td>27984722</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSDARG00000100083</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>havana</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSDARE00001151908</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1:27984392-27984722</td>\n",
       "      <td>330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>27984815</td>\n",
       "      <td>27984885</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSDARG00000100083</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>havana</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSDARE00001193383</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1:27984815-27984885</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>27993184</td>\n",
       "      <td>27993255</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSDARG00000100083</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>havana</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSDARE00001181473</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1:27993184-27993255</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1161860</th>\n",
       "      <td>MT</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15232</td>\n",
       "      <td>15301</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSDARG00000083312</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MT:15232-15301</td>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1161861</th>\n",
       "      <td>MT</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>15232</td>\n",
       "      <td>15301</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSDARG00000083312</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSDARE00000882905</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MT:15232-15301</td>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1161862</th>\n",
       "      <td>MT</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>16526</td>\n",
       "      <td>16596</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSDARG00000081475</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MT:16526-16596</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1161863</th>\n",
       "      <td>MT</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>16526</td>\n",
       "      <td>16596</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSDARG00000081475</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MT:16526-16596</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1161864</th>\n",
       "      <td>MT</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>16526</td>\n",
       "      <td>16596</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSDARG00000081475</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSDARE00000882281</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MT:16526-16596</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1161865 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome          Source     Feature     Start       End Score  \\\n",
       "0                1  ensembl_havana        gene  27977296  28020042     .   \n",
       "1                1          havana  transcript  27984392  27995611     .   \n",
       "2                1          havana        exon  27984392  27984722     .   \n",
       "3                1          havana        exon  27984815  27984885     .   \n",
       "4                1          havana        exon  27993184  27993255     .   \n",
       "...            ...             ...         ...       ...       ...   ...   \n",
       "1161860         MT          RefSeq  transcript     15232     15301     .   \n",
       "1161861         MT          RefSeq        exon     15232     15301     .   \n",
       "1161862         MT          RefSeq        gene     16526     16596     .   \n",
       "1161863         MT          RefSeq  transcript     16526     16596     .   \n",
       "1161864         MT          RefSeq        exon     16526     16596     .   \n",
       "\n",
       "        Strand Frame             gene_id gene_version  ... transcript_source  \\\n",
       "0            +     .  ENSDARG00000100083            2  ...               NaN   \n",
       "1            +     .  ENSDARG00000100083            2  ...            havana   \n",
       "2            +     .  ENSDARG00000100083            2  ...            havana   \n",
       "3            +     .  ENSDARG00000100083            2  ...            havana   \n",
       "4            +     .  ENSDARG00000100083            2  ...            havana   \n",
       "...        ...   ...                 ...          ...  ...               ...   \n",
       "1161860      -     .  ENSDARG00000083312            3  ...            RefSeq   \n",
       "1161861      -     .  ENSDARG00000083312            3  ...            RefSeq   \n",
       "1161862      -     .  ENSDARG00000081475            3  ...               NaN   \n",
       "1161863      -     .  ENSDARG00000081475            3  ...            RefSeq   \n",
       "1161864      -     .  ENSDARG00000081475            3  ...            RefSeq   \n",
       "\n",
       "           transcript_biotype exon_number             exon_id exon_version  \\\n",
       "0                         NaN         NaN                 NaN          NaN   \n",
       "1        processed_transcript         NaN                 NaN          NaN   \n",
       "2        processed_transcript           1  ENSDARE00001151908            1   \n",
       "3        processed_transcript           2  ENSDARE00001193383            1   \n",
       "4        processed_transcript           3  ENSDARE00001181473            1   \n",
       "...                       ...         ...                 ...          ...   \n",
       "1161860               Mt_tRNA         NaN                 NaN          NaN   \n",
       "1161861               Mt_tRNA           1  ENSDARE00000882905            3   \n",
       "1161862                   NaN         NaN                 NaN          NaN   \n",
       "1161863               Mt_tRNA         NaN                 NaN          NaN   \n",
       "1161864               Mt_tRNA           1  ENSDARE00000882281            3   \n",
       "\n",
       "        protein_id protein_version  tag             interval Length  \n",
       "0              NaN             NaN  NaN  1:27977296-28020042  42746  \n",
       "1              NaN             NaN  NaN  1:27984392-27995611  11219  \n",
       "2              NaN             NaN  NaN  1:27984392-27984722    330  \n",
       "3              NaN             NaN  NaN  1:27984815-27984885     70  \n",
       "4              NaN             NaN  NaN  1:27993184-27993255     71  \n",
       "...            ...             ...  ...                  ...    ...  \n",
       "1161860        NaN             NaN  NaN       MT:15232-15301     69  \n",
       "1161861        NaN             NaN  NaN       MT:15232-15301     69  \n",
       "1161862        NaN             NaN  NaN       MT:16526-16596     70  \n",
       "1161863        NaN             NaN  NaN       MT:16526-16596     70  \n",
       "1161864        NaN             NaN  NaN       MT:16526-16596     70  \n",
       "\n",
       "[1161865 rows x 26 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a pyranges of gene annotation used for counting expression (gtf)\n",
    "GTF = pr.read_gtf(f'{commonsdir}/annotations/Danio_rerio.GRCz11.104.gtf.gz')\n",
    "GTF = add_interval_and_length(GTF)\n",
    "GTF = GTF.as_df()\n",
    "GTF\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "3a7f42fc-4387-41fd-a8af-de1eb1843d02",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_version</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_source</th>\n",
       "      <th>transcript_biotype</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_version</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>protein_version</th>\n",
       "      <th>tag</th>\n",
       "      <th>interval</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>28018748</td>\n",
       "      <td>28019225</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSDARG00000100083</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1:28018748-28019225</td>\n",
       "      <td>477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>27984863</td>\n",
       "      <td>27984885</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSDARG00000100083</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>havana</td>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1:27984863-27984885</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>27993184</td>\n",
       "      <td>27993255</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSDARG00000100083</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>havana</td>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1:27993184-27993255</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>27994791</td>\n",
       "      <td>27994845</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSDARG00000100083</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>havana</td>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1:27994791-27994845</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>27994938</td>\n",
       "      <td>27994967</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSDARG00000100083</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>havana</td>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1:27994938-27994967</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1161620</th>\n",
       "      <td>KZ116044.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>13791</td>\n",
       "      <td>14027</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSDARG00000115450</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KZ116044.1:13791-14027</td>\n",
       "      <td>236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1161633</th>\n",
       "      <td>KZ116046.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>20</td>\n",
       "      <td>2861</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSDARG00000113284</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KZ116046.1:20-2861</td>\n",
       "      <td>2841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1161667</th>\n",
       "      <td>KZ116055.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>290</td>\n",
       "      <td>516</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSDARG00000109919</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KZ116055.1:290-516</td>\n",
       "      <td>226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1161679</th>\n",
       "      <td>KZ116064.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>82791</td>\n",
       "      <td>82859</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSDARG00000114977</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KZ116064.1:82791-82859</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1161714</th>\n",
       "      <td>KZ116065.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>20388</td>\n",
       "      <td>20619</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSDARG00000102423</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KZ116065.1:20388-20619</td>\n",
       "      <td>231</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34772 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Chromosome          Source          Feature     Start       End  \\\n",
       "39                1  ensembl_havana  three_prime_utr  28018748  28019225   \n",
       "56                1          havana  three_prime_utr  27984863  27984885   \n",
       "57                1          havana  three_prime_utr  27993184  27993255   \n",
       "58                1          havana  three_prime_utr  27994791  27994845   \n",
       "59                1          havana  three_prime_utr  27994938  27994967   \n",
       "...             ...             ...              ...       ...       ...   \n",
       "1161620  KZ116044.1         ensembl  three_prime_utr     13791     14027   \n",
       "1161633  KZ116046.1         ensembl  three_prime_utr        20      2861   \n",
       "1161667  KZ116055.1         ensembl  three_prime_utr       290       516   \n",
       "1161679  KZ116064.1         ensembl  three_prime_utr     82791     82859   \n",
       "1161714  KZ116065.1         ensembl  three_prime_utr     20388     20619   \n",
       "\n",
       "        Score Strand Frame             gene_id gene_version  ...  \\\n",
       "39          .      +     .  ENSDARG00000100083            2  ...   \n",
       "56          .      +     .  ENSDARG00000100083            2  ...   \n",
       "57          .      +     .  ENSDARG00000100083            2  ...   \n",
       "58          .      +     .  ENSDARG00000100083            2  ...   \n",
       "59          .      +     .  ENSDARG00000100083            2  ...   \n",
       "...       ...    ...   ...                 ...          ...  ...   \n",
       "1161620     .      +     .  ENSDARG00000115450            1  ...   \n",
       "1161633     .      -     .  ENSDARG00000113284            1  ...   \n",
       "1161667     .      -     .  ENSDARG00000109919            1  ...   \n",
       "1161679     .      +     .  ENSDARG00000114977            1  ...   \n",
       "1161714     .      +     .  ENSDARG00000102423            2  ...   \n",
       "\n",
       "        transcript_source       transcript_biotype exon_number exon_id  \\\n",
       "39         ensembl_havana           protein_coding         NaN     NaN   \n",
       "56                 havana  nonsense_mediated_decay         NaN     NaN   \n",
       "57                 havana  nonsense_mediated_decay         NaN     NaN   \n",
       "58                 havana  nonsense_mediated_decay         NaN     NaN   \n",
       "59                 havana  nonsense_mediated_decay         NaN     NaN   \n",
       "...                   ...                      ...         ...     ...   \n",
       "1161620           ensembl           protein_coding         NaN     NaN   \n",
       "1161633           ensembl           protein_coding         NaN     NaN   \n",
       "1161667           ensembl           protein_coding         NaN     NaN   \n",
       "1161679           ensembl           protein_coding         NaN     NaN   \n",
       "1161714           ensembl           protein_coding         NaN     NaN   \n",
       "\n",
       "        exon_version protein_id protein_version  tag                interval  \\\n",
       "39               NaN        NaN             NaN  NaN     1:28018748-28019225   \n",
       "56               NaN        NaN             NaN  NaN     1:27984863-27984885   \n",
       "57               NaN        NaN             NaN  NaN     1:27993184-27993255   \n",
       "58               NaN        NaN             NaN  NaN     1:27994791-27994845   \n",
       "59               NaN        NaN             NaN  NaN     1:27994938-27994967   \n",
       "...              ...        ...             ...  ...                     ...   \n",
       "1161620          NaN        NaN             NaN  NaN  KZ116044.1:13791-14027   \n",
       "1161633          NaN        NaN             NaN  NaN      KZ116046.1:20-2861   \n",
       "1161667          NaN        NaN             NaN  NaN      KZ116055.1:290-516   \n",
       "1161679          NaN        NaN             NaN  NaN  KZ116064.1:82791-82859   \n",
       "1161714          NaN        NaN             NaN  NaN  KZ116065.1:20388-20619   \n",
       "\n",
       "        Length  \n",
       "39         477  \n",
       "56          22  \n",
       "57          71  \n",
       "58          54  \n",
       "59          29  \n",
       "...        ...  \n",
       "1161620    236  \n",
       "1161633   2841  \n",
       "1161667    226  \n",
       "1161679     68  \n",
       "1161714    231  \n",
       "\n",
       "[34772 rows x 26 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Subset to 3'UTRs\n",
    "GTF = GTF[GTF.Feature == 'three_prime_utr']\n",
    "GTF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "718fe6c7-b47d-493c-9adb-1bb005a8eb85",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/scott/scripts/scmultiome_curbos_scripts/multiome/lib/python3.11/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    }
   ],
   "source": [
    "# Save 3'UTRs in a bed file for use with bedtools\n",
    "pr.PyRanges(GTF[['Chromosome', 'Start', 'End', 'interval', 'Score', 'Strand']]).to_bed(\n",
    "    path = f'{datadir}/GRCz11_104_3_prime_UTR.bed', \n",
    "    keep=True, \n",
    "    compression='infer', \n",
    "    chain=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "b59ef3f2-9956-48a6-ace8-b543126e4134",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract sequences associated with 3' UTRs and save to fasta\n",
    "a = pybedtools.BedTool(f'{datadir}/GRCz11_104_3_prime_UTR.bed')\n",
    "\n",
    "fasta = pybedtools.example_filename(\"/plus/scratch/sai/seqs/danrer_genome_all_ensembl_grcz11.fa\")\n",
    "\n",
    "a = a.sequence(fi=fasta)\n",
    "\n",
    "b = a.save_seqs(f'{datadir}/GRCz11_104_3_prime_UTR.fa')\n",
    "assert open(b.fn).read() == open(a.fn).read()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "356f9459-740a-464d-9c17-ec092d0a2879",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create dataframe of sequences\n",
    "from Bio import SeqIO\n",
    "\n",
    "with open(f'{datadir}/GRCz11_104_3_prime_UTR.fa') as fasta_file:  # Will close handle cleanly\n",
    "    intervals = []\n",
    "    seqs = []\n",
    "    for seq_record in SeqIO.parse(fasta_file, 'fasta'):  # (generator)\n",
    "        intervals.append(seq_record.id)\n",
    "        seqs.append(str(seq_record.seq))\n",
    "\n",
    "interval_seqs = pd.DataFrame({'3_prime_UTR_sequence': seqs}, index=intervals)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "b5ec819c-d0ea-46af-bdca-5b80567f1284",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Join GTF and sequences dataframe\n",
    "GTF.index = GTF.interval\n",
    "GTF = pd.concat([GTF, interval_seqs], axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "1c6a4b41-3407-4234-9b4a-bec14e7a1800",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_699797/2106026537.py:9: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!\n",
      "You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.\n",
      "A typical example is when you are setting values in a column of a DataFrame, like:\n",
      "\n",
      "df[\"col\"][row_indexer] = value\n",
      "\n",
      "Use `df.loc[row_indexer, \"col\"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "\n",
      "  GTF['mir430_motif'][GTF['mir430_motif6']] = 6\n",
      "/tmp/ipykernel_699797/2106026537.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  GTF['mir430_motif'][GTF['mir430_motif6']] = 6\n",
      "/tmp/ipykernel_699797/2106026537.py:9: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '6' has dtype incompatible with bool, please explicitly cast to a compatible dtype first.\n",
      "  GTF['mir430_motif'][GTF['mir430_motif6']] = 6\n"
     ]
    }
   ],
   "source": [
    "# Identify those with mir430 sites (6, 7, 8 mer)\n",
    "GTF['mir430_motif8'] = GTF['3_prime_UTR_sequence'].str.contains('AGCACTTA')\n",
    "GTF['mir430_motif7'] = GTF['3_prime_UTR_sequence'].str.contains('AGCACTT|GCACTTA')\n",
    "GTF['mir430_motif6'] = GTF['3_prime_UTR_sequence'].str.contains('AGCACT|GCACTT')\n",
    "GTF['mir430_motif_count'] = GTF['3_prime_UTR_sequence'].str.count('GCACTT')\n",
    "GTF['mir430_motif_seed'] = GTF['3_prime_UTR_sequence'].str.contains('GCACTT')\n",
    "\n",
    "GTF['mir430_motif'] = GTF['mir430_motif6']\n",
    "GTF['mir430_motif'][GTF['mir430_motif6']] = 6\n",
    "GTF['mir430_motif'][GTF['mir430_motif7']] = 7\n",
    "GTF['mir430_motif'][GTF['mir430_motif8']] = 8\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "d8ccc454-5b32-479d-9d06-29f86c7d8f0e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_699797/2763247141.py:8: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!\n",
      "You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.\n",
      "A typical example is when you are setting values in a column of a DataFrame, like:\n",
      "\n",
      "df[\"col\"][row_indexer] = value\n",
      "\n",
      "Use `df.loc[row_indexer, \"col\"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "\n",
      "  GTF['RC_mir430_motif'][GTF['RC_mir430_motif6']] = 6\n",
      "/tmp/ipykernel_699797/2763247141.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  GTF['RC_mir430_motif'][GTF['RC_mir430_motif6']] = 6\n",
      "/tmp/ipykernel_699797/2763247141.py:8: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '6' has dtype incompatible with bool, please explicitly cast to a compatible dtype first.\n",
      "  GTF['RC_mir430_motif'][GTF['RC_mir430_motif6']] = 6\n"
     ]
    }
   ],
   "source": [
    "# Identify those with reverse complement mir430 sites (6, 7, 8 mer)\n",
    "GTF['RC_mir430_motif8'] = GTF['3_prime_UTR_sequence'].str.contains('TAAGTGCT')\n",
    "GTF['RC_mir430_motif7'] = GTF['3_prime_UTR_sequence'].str.contains('TAAGTGC|AAGTGCT')\n",
    "GTF['RC_mir430_motif6'] = GTF['3_prime_UTR_sequence'].str.contains('TAAGTG|AAGTGC|AGTGCT')\n",
    "GTF['RC_mir430_motif_seed'] = GTF['3_prime_UTR_sequence'].str.contains('AAGTGC')\n",
    "\n",
    "GTF['RC_mir430_motif'] = GTF['RC_mir430_motif6']\n",
    "GTF['RC_mir430_motif'][GTF['RC_mir430_motif6']] = 6\n",
    "GTF['RC_mir430_motif'][GTF['RC_mir430_motif7']] = 7\n",
    "GTF['RC_mir430_motif'][GTF['RC_mir430_motif8']] = 8\n",
    "\n",
    "GTF['mir430_F_RC_motif'] = (GTF['RC_mir430_motif'] != False) | (GTF['mir430_motif'] != False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "2cc839c3-8768-4595-b0fe-e6a9270d54ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save 3'UTR annotation\n",
    "GTF.to_csv(f'{datadir}/GRCz11_104_3_prime_UTR.csv', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8111cdb-912d-41be-8a55-8190493e8b04",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "multiome",
   "language": "python",
   "name": "multiome"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
